### ISSUE TITLE
KeypointRCNN and RoIHeads question from a Bachelor student

### ISSUE THREAD

**Issue by SergiHernandez** [pmeier/pmeier#13]
Hello @pmeier ,

I hope this issue finds you well.

I've seen you actively contribute to the torchvision/detection project, so I wanted to reach for help.

For my bachelor thesis I need to change the KeypointRCNNPredictor module by a Graph Convolutional Network (GCN).

The standard keypoint predictor takes a feature for each of the B proposed bounding boxes (B×512×14×14) and applies transposed convolutions and interpolation to upscale the feature maps. So the outputs are heatmaps of logits for each keypoint for each box (B×P×56×56). These logits are then passed to the keypoint_loss function that is implemented within the RoIHeads source code. This loss takes the target keypoints, transforms them into heatmaps too, and then computes the CrossEntropyLoss between the target heatmaps and the logit heatmaps outputed by the KeypointRCNNPredictor.

In the case of the GCN, it takes also feature maps, but outputs a list of keypoint coordinates. So for every bounding box B, it outputs the x,y coordinates of the P keypoints within that box (B×P×2).

If I change the standard keypoint predictor by the GCN, I have two options:
- Option 1. (hard to implement but more correct from my point on view) Overwriting the keypoint loss function for it to be a MSELoss, or some other distance loss, to take the loss as the distance between the predicted points by the GCN and the target coordinates. This also requires to implement again the keypoint inference function within RoIHeads because the forward takes different paths in train and eval mode. And probably some other auxiliary functions need some work too.
- Option 2. (very easy if doable) Convert the coordinates output of the GCN into a heatmap and let the losses and inference functions work as usual with heatmaps. The problem with this is that I am not sure about how to build a heatmap (kind of a parabola centered at the predicted keypoint), such that it works with the backpropagation.

I would highly appreciate your guidance and your suggestions. I am a month away from delivering the thesis and I just need this last step to finish it.

Many thanks for your help!!
Sergi

PD: if it is more convenient for you, please contact me at sergi.hernandez01@estudiant.upf.edu

**Comment by pmeier** [comment]
Hey Sergi :wave: Just to make sure we are talking about the same thing: this is the [loss function](https://github.com/pytorch/vision/blob/947ae1dc71867f28021d5bc0ff3a19c249236e2a/torchvision/models/detection/roi_heads.py#L298) you want to swap and it is called from [here](https://github.com/pytorch/vision/blob/947ae1dc71867f28021d5bc0ff3a19c249236e2a/torchvision/models/detection/roi_heads.py#L860-L862), correct?

IIUC your proposal above correctly, everything stays the same, but you actually want to just have different code there for your use case, correct?

**Comment by SergiHernandez** [comment]
Thanks for answering!

Exactly. The GCN returns the keypoints as x,y coordinates. So appart from swapping the [loss function](https://github.com/pytorch/vision/blob/947ae1dc71867f28021d5bc0ff3a19c249236e2a/torchvision/models/detection/roi_heads.py#L298) to use a MSELoss instead of the CrossEntrpoy, I need to change the [keypoint inference function](https://github.com/pytorch/vision/blob/947ae1dc71867f28021d5bc0ff3a19c249236e2a/torchvision/models/detection/roi_heads.py#L329), that takes care of transforming from relative positions in the bounding box to absolute positions in the whole image.

Moreover, I've seen that after the RoI Heads is called, GeneralizedRCNN calls a [postprocess function](https://github.com/pytorch/vision/blob/947ae1dc71867f28021d5bc0ff3a19c249236e2a/torchvision/models/detection/transform.py#L257) that also does some resizing, but I don't get to understand where this last resizing comes from.

I don't know if more functions should be changed. I am afraid I am missing something in the process, because the predictions after doing some changes are very bad while with the standard KeypointRCNN are very good.

And yes, the thesis experiment involves only replacing the KeypointRCNNPredictor module by a GCN for my use case, the rest of the network stays the same if possible.
